server/src/main/java/io/druid/server/ClientInfoResource.java

```diff
@@ -123,33 +119,12 @@ public ClientInfoResource(
   @Produces(MediaType.APPLICATION_JSON)
   public Iterable<String> getDataSources(@Context final HttpServletRequest request)
   {
-    if (authConfig.isEnabled()) {
-      // This is an experimental feature, see - https://github.com/druid-io/druid/pull/2424
-      final Map<Pair<Resource, Action>, Access> resourceAccessMap = new HashMap<>();
-      final AuthorizationInfo authorizationInfo = (AuthorizationInfo) request.getAttribute(AuthConfig.DRUID_AUTH_TOKEN);
-      return Collections2.filter(
-          getSegmentsForDatasources().keySet(),
-          new Predicate<String>()
-          {
-            @Override
-            public boolean apply(String input)
-            {
-              Resource resource = new Resource(input, ResourceType.DATASOURCE);
-              Action action = Action.READ;
-              Pair<Resource, Action> key = new Pair<>(resource, action);
-              if (resourceAccessMap.containsKey(key)) {
-                return resourceAccessMap.get(key).isAllowed();
-              } else {
-                Access access = authorizationInfo.isAuthorized(key.lhs, key.rhs);
-                resourceAccessMap.put(key, access);
-                return access.isAllowed();
-              }
-            }
-          }
-      );
-    } else {
-      return getSegmentsForDatasources().keySet();
-    }
+    return AuthorizationUtils.filterAuthorizedResources(
+        request,
+        getSegmentsForDatasources().keySet(),
+        AuthorizationUtils.DATASOURCE_READ_RA_GENERATOR,
+        authorizerMapper
```

Comment:
A thought: make the AuthorizationUtils utility methods into methods on authorizerMapper?

Reply:
Like `authorizerMapper.filter(request, getSegmentsForDatasources().keySet(), AuthorizationUtils.DATASOURCE_READ_RA_GENERATOR);`